src/zarr/storage/object_store.py

```diff
@@ -0,0 +1,306 @@
+
+            return prototype.buffer.from_bytes(await resp.bytes_async())
+        else:
+            raise ValueError(f"Unexpected input to `get`: {start=}, {end=}")
+
+    async def get_partial_values(
+        self,
+        prototype: BufferPrototype,
+        key_ranges: Iterable[tuple[str, ByteRangeRequest]],
+    ) -> list[Buffer | None]:
+        return await _get_partial_values(self.store, prototype=prototype, key_ranges=key_ranges)
+
+    async def exists(self, key: str) -> bool:
+        try:
+            await obs.head_async(self.store, key)
+        except FileNotFoundError:
+            return False
+        else:
+            return True
+
+    @property
+    def supports_writes(self) -> bool:
+        return True
+
+    async def set(self, key: str, value: Buffer) -> None:
+        buf = value.to_bytes()
+        await obs.put_async(self.store, key, buf)
+
+    async def set_if_not_exists(self, key: str, value: Buffer) -> None:
+        buf = value.to_bytes()
+        await obs.put_async(self.store, key, buf, mode="create")
+
+    @property
+    def supports_deletes(self) -> bool:
+        return True
+
+    async def delete(self, key: str) -> None:
+        await obs.delete_async(self.store, key)
+
+    @property
+    def supports_partial_writes(self) -> bool:
+        return False
+
+    async def set_partial_values(
+        self, key_start_values: Iterable[tuple[str, int, BytesLike]]
+    ) -> None:
+        raise NotImplementedError
+
+    @property
+    def supports_listing(self) -> bool:
+        return True
+
+    def list(self) -> AsyncGenerator[str, None]:
+        objects: ListStream[list[ObjectMeta]] = obs.list(self.store)
+        return _transform_list(objects)
+
+    def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
+        objects: ListStream[list[ObjectMeta]] = obs.list(self.store, prefix=prefix)
+        return _transform_list(objects)
+
+    def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
+        objects: ListStream[list[ObjectMeta]] = obs.list(self.store, prefix=prefix)
+        return _transform_list_dir(objects, prefix)
+
+
+async def _transform_list(
+    list_stream: AsyncGenerator[list[ObjectMeta], None],
+) -> AsyncGenerator[str, None]:
+    async for batch in list_stream:
+        for item in batch:
+            yield item["path"]
+
+
+async def _transform_list_dir(
+    list_stream: AsyncGenerator[list[ObjectMeta], None], prefix: str
+) -> AsyncGenerator[str, None]:
+    # We assume that the underlying object-store implementation correctly handles the
+    # prefix, so we don't double-check that the returned results actually start with the
+    # given prefix.
+    prefix_len = len(prefix)
+    async for batch in list_stream:
+        for item in batch:
+            # Yield this item if "/" does not exist after the prefix.
+            if "/" not in item["path"][prefix_len:]:
+                yield item["path"]
```

Comment:
It was brought to my attention in https://github.com/developmentseed/obstore/issues/101 that in the current `object_store` implementation a final `/` is stripped from directories. So this `_transform_list_dir` will include the names of any sub directories.

Does zarr expect zero-byte files in normal usage? Or we could assume that a zero-byte file may be a directory.

Reply:
fsspec has struggled with this for the longest time. My recommendation: you won't come up with a convention that matches the expectations of all users and backends, so choose something you're happy with and maintain it consistently. As long as the documentation is clear about how the convention works.

Reply:
Thanks, that's useful background. What does that mean for Zarr? What does Zarr expect? Should we document behavior at the `Store` level for what all stores should implement? So at least the fsspec-based and obstore-based stores could have the same behavior

Reply:
I thought that listing and caring about directories versus files is something that only matters for V2, no?

Reply:
I'm not intricately familiar with Zarr v3 but there are still `Store` methods for `list`, `list_prefix` and `list_dir`. Will it give incorrect behavior if one of the listed files is a directory prefix with no content?

Reply:
I am unsure, but I would have thought not.

Reply:
Well, good for this backend to be considered experimental for now 🙂